### PR TITLE
[#147445891] Compose integration tests

### PR DESCRIPTION
**Pull request by LeePorte**
## What 

This PR adds an integration configuration and version file for running integration tests on PRs via Concourse in the CI environment. In addition pipecleaner has had to have some additional work undertaken in order to handle external task files and resources loaded in the main pipeline and then consumed in the external task file.

## How to review 

1. https://github.com/alphagov/paas-credentials/pull/82 needs to be merged first
1. Create a dev build concourse via paas-bootstrap
1. Upload compose secrets with `make dev upload-compose-secrets`
1. Push the pipelines from this branch `make dev pipelines` for setup, `make dev integration-test-pipelines` for the integration pipelines, these include the new test and tag pipeline.
1. Run the Setup pipeline, this will generate a SSH keypair
1. Login to Github with the CI user (in creds store) and add the generated pub key to the user
1. Run the Compose broker and RDS broker pipelines
1. Delete pub key from CI Github user

## Merging

1. `make ci upload-compose-secrets` must be performed by someone with access to the high creds store.
1. Remove the tmp commit that points at the compose_integration_branch
1. Following this, the merge process is as usual.

## Who can review 

Not @combor or @LeePorte or @paroxp or @dcarley

### CODE REVIEW

File: scripts/upload-compose-secrets.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        COMPOSE_ACCESS_TOKEN=$(pass "compose/${AWS_ACCOUNT}/access_token")
+fi
+
+
+SECRETS=$(mktemp secrets.yml.XXXXXX)
+trap 'rm  "${SECRETS}"' EXIT
+
+cat > "${SECRETS}" << EOF
+---
+compose_access_token: ${COMPOSE_ACCESS_TOKEN}
```
Comment (LeePorte) [inline scripts/upload-compose-secrets.sh]: I'm inclined to stick with it as it is the pattern we have for other secrets in the repo and will prove useful if we need to pass in additional secrets.

File: .travis.yml [hunk 2]
```diff
@@ -19,7 +19,7 @@ before_install:
     unzip -o terraform_${TF_VERSION}_linux_amd64.zip -d ~/bin
     rm terraform_${TF_VERSION}_linux_amd64.zip
   - pip install --user yamllint
-  - eval "$(gimme 1.6)"
+  - eval "$(gimme 1.8)"
```
Comment (LeePorte) [inline .travis.yml]: Lint errors occurred using 1.6, in relation to an import. Upgrading to 1.8 fixed these.

Comment (dcarley) [inline .travis.yml]: Do you know what the errors were? Can you split it out to a separate commit with the details?

File: Makefile [hunk 3]
```diff
@@ -47,14 +52,13 @@ boshrelease-pipelines: ## Upload boshrelease pipelines to concourse
 app-deployment-pipelines: ## Upload app deployment pipelines to concourse
 	@scripts/build-app-deployment-pipelines.sh
 
-.PHONY: integration-test-pipelines
-integration-test-pipelines: ## Upload integration test pipelines to concourse
-	@scripts/integration-test-pipelines.sh
+.PHONY: git-release-pipelines
```
Comment (dcarley) [inline Makefile]: I think we should use the existing `integration-test-pipelines` target and extend the `pipelines/integration-test.yml` pipeline to do tagging because it will make the changes easier to read, rather than removing everything and adding new.

I'm indifferent about whether we rename it to something like `integration-test-tag` to indicate that it now does tagging. If we do rename though, it would be good to do that in a separate commit, again so that history is easier to read.

Could you also squash the "combine pipelines for PR feedback" commit when you do so?

**Comment by LeePorte** [comment]
@dcarley tweaks made :)

File: pipelines/setup.yml [hunk 4]
```diff
@@ -141,11 +157,61 @@ jobs:
...
+          run:
+            path: sh
+            args:
+              - -e
+              - -c
+              - |
+                if [ -s ssh-private-key/ci_build_tag_key ] ; then
+                  echo "Deployments private key non-zero size, skipping generation..."
+                  echo "Key uploads will fail and this is OK as no new keys have been generated."
+                  exit 0
```
Comment (dcarley) [inline pipelines/setup.yml]: Can this job output the public key so that you can copy/paste it without manually looking in the bucket?

It would be good to do this when the key is first generated below too.

File: pipelines/integration-test.yml [hunk 5]
```diff
@@ -40,3 +66,27 @@ jobs:
...
+    serial: true
+    plan:
+    - aggregate:
+      - get: release-repository
+        trigger: true
+      - get: resource-version
+        params:
+          bump: minor
+      - task: run-tests
+        file: pr/ci/integration.yml
```
Comment (dcarley) [inline pipelines/integration-test.yml]: I get this error from the `run-tests` task:
```
unknown artifact source: pr
```

I think `pr` needs to be renamed to `release-repository` to match the resource name.

File: scripts/integration-test-pipelines.sh [hunk 6]
```diff
@@ -9,24 +9,44 @@ $("${SCRIPTS_DIR}/environment.sh")
...
   bash "${SCRIPTS_DIR}/deploy-pipeline.sh" \
     "${pipeline_name}" \
     "${PIPELINES_DIR}/integration-test.yml" \
-    <(generate_vars_file)
+    <(generate_vars_file "$@") 
+
 }
 
-setup_test_pipeline rds-broker alphagov/paas-rds-broker
+setup_test_pipeline rds-broker alphagov paas-rds-broker master
```
Comment (LeePorte) [inline scripts/integration-test-pipelines.sh]: Yes we will

**Comment by dcarley** [comment]
I've pushed some commits to:

- fix the pipecleaner errors, as separate commits describing the problems
- rename the resource to `repo` in paas-compose-broker and paas-rds-broker, as a fixup commit that can be squashed if you're happy

I haven't been able to test paas-rds-broker because I found that you can only an SSH public key as a deploy key to a single repo. So that might force our decision about user management and blocks this story a bit. Let's talk about it on Monday.

Someone else should probably review the story afterwards now that I've contributed to it.

File: scripts/upload-compose-secrets.sh [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+set -eu
+
+if [ "${AWS_ACCOUNT}" = "dev" ]
+   then
+        export PASSWORD_STORE_DIR="${HOME}"/.paas-pass
+        COMPOSE_ACCESS_TOKEN=$(pass "compose/integration_tests/access_token")
+   else
+        export COMPOSE_PASSWORD_STORE_DIR="${HOME}"/.paas-pass-high
+        COMPOSE_ACCESS_TOKEN=$(pass "compose/${AWS_ACCOUNT}/access_token")
```
Comment (alext) [inline scripts/upload-compose-secrets.sh]: This should be using the integration_tests token in both environments. The CI api keys should only be used by the CI CF deployment (they belong to the CI PaaS-CF user).

File: scripts/integration-test-pipelines.sh [hunk 8]
```diff
@@ -9,24 +9,44 @@ $("${SCRIPTS_DIR}/environment.sh")
...
 }
 
 setup_test_pipeline() {
   pipeline_name="$1"
-  github_repo="$2"
+  version_file="version"
+  organisation="$2"
+  repository="$3"
+  tag_branch="$4"
+  secrets_file="${5:-no-secrets-needed}"
```
Comment (LeePorte) [inline scripts/integration-test-pipelines.sh]: Added in https://github.com/alphagov/paas-release-ci/pull/30/files#diff-203673b1eaa54b56dcea63cd041ec574R119

File: scripts/integration-test-pipelines.sh [hunk 9]
```diff
@@ -9,24 +9,44 @@ $("${SCRIPTS_DIR}/environment.sh")
...
+branch_name: ${BRANCH:-master}
+organisation: ${organisation}
+repository: ${repository}
+github_repo: ${organisation}/${repository}
+github_repo_uri: git@github.com:${organisation}/${repository}.git
+tag_branch: ${tag_branch}
+version_file: ${version_file}
+secrets_file: ${secrets_file}
 EOF
+echo -e "tagging_key: |\n  ${SSH_KEY//$'\n'/$'\n'  }"
```
Comment (alext) [inline scripts/integration-test-pipelines.sh]: Now that I read this more fully, I realise what this is doing. It's meant to be here because this forms part of the YAML.

File: scripts/upload-compose-secrets.sh [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+set -eu
+
+if [ "${AWS_ACCOUNT}" = "dev" ]
+   then
+        export PASSWORD_STORE_DIR="${HOME}"/.paas-pass
+        COMPOSE_ACCESS_TOKEN=$(pass "compose/integration_tests/access_token")
+   else
+        export COMPOSE_PASSWORD_STORE_DIR="${HOME}"/.paas-pass-high
```
Comment (dcarley) [inline scripts/upload-compose-secrets.sh]: There's some previous context where we've moved them about previously in:

- https://github.com/alphagov/paas-release-ci/pull/30#discussion_r128785439
- https://github.com/alphagov/paas-release-ci/pull/30#discussion_r128794612
